### PR TITLE
fix: use hidden to mask seed words

### LINKED ISSUE #4861: `Mnemonic::to_mnemonic` should wrap the seed words in Zeroizing
https://github.com/tari-project/tari/blob/5b0eb0459c7d29a25339c22a289153d27d57388e/base_layer/key_manager/src/mnemonic.rs#L254

The TariSeedWords ffi struct may need to be updated, but I don't think this will cause a breaking change as it will not change the underlying representation

### PR DESCRIPTION
**Pull request by brianp**
Description
---
Don't let seed words leak into logfiles or other places. Mask them and only reveal them when specifically required.
I didn't use zeroize because I know we've had some issues with it and `Hidden` seemed like it would do the job well enough. Happy to change it if there are any strong preferences.

Motivation and Context
---
Prevent leaked secrets.

How Has This Been Tested?
---
Manually checking to ensure seed words can be returned and written to disk without a problem.

Closes: #4861

### CODE REVIEW

**Review by stringhandler — DISMISSED** [review]

**Comment by stringhandler** [comment]
utACK

**Review by stringhandler — DISMISSED** [review]

File: base_layer/key_manager/src/cipher_seed.rs [hunk 1]
```diff
@@ -34,6 +34,7 @@ use rand::{rngs::OsRng, RngCore};
 use serde::{Deserialize, Serialize};
 use subtle::ConstantTimeEq;
 use tari_crypto::hash::blake2::Blake256;
+use tari_utilities::{ByteArray, Hidden};
```
Comment (stringhandler) [inline base_layer/key_manager/src/cipher_seed.rs]:
```suggestion
use tari_utilities::Hidden;
```

**Comment by AaronFeickert** [comment]
What issues specifically are of concern with `zeroize`? It is true that [moves and copies can be problematic](https://docs.rs/zeroize/latest/zeroize/#stackheap-zeroing-notes), but `Hidden` appears to do nothing for leaving in-memory traces of sensitive data.

**Comment by sdbondi** [comment]
Agree with @AaronFeickert that we should be using the Zeroizing wrapper. The problems with moving are only tricky for data that lives on the stack. With `Vec<String>` all sensitive data is heap-allocated Zeroizing will zeroize [each element in the vec](https://github.com/RustCrypto/utils/blob/b898588b13e8005cd8860089f108bed09f86bd87/zeroize/src/lib.rs#L541).

**Comment by brianp** [comment]
> What issues specifically are of concern with zeroize?

I believe the problem wasn't how it did anything but how it's being loaded and places it caused problems across crates.

I'm happy to move it to zeroize right now as I hadn't realized that key difference. As that's an important feature for us, maybe we should expand `Hidden` at some point.

**Comment by stringhandler** [comment]
Maybe `Hidden` should implement `Zeroize`?

**Comment by sdbondi** [comment]
SGTM, Hidden makes sense for seed words, in that you don't want any derive Debug accidents and `Hidden<T>` by definition contains sensitive data so should zeroize on drop for `T: Zeroize`.

**Comment by CjS77** [comment]
The convo is inferring a lot of composability. 

Seed words _are_ hidden, seed words _are_ zeroizable etc.

This suggests that there's value in defining a `SeedWord` type:

`#[derive(zeroize)] pub struct SeedWord(Hidden<String>)`

**Comment by CjS77** [comment]
> SGTM, Hidden makes sense for seed words, in that you don't want any derive Debug accidents and `Hidden<T>` by definition contains sensitive data so should zeroize on drop for `T: Zeroize`.

Also good. Perhaps the Zeroize trait bound should be compulsory, i.e. defined on the struct rather than the impl

**Comment by AaronFeickert** [comment]
So would the overall idea be to add a `Zeroize` trait bound to `Hidden`, and then create the new `SeedWord` type that zeroizes on drop? It looks like `Hidden` implements `Copy`, so as I understand it you couldn't zeroize it on drop.

**Comment by brianp** [comment]
`Hidden` isn't widely used yet, so if the zeroized version of it becomes the standard I think that's fine if that's the expectation of the struct, and it's inability to be copied.

Looking closer everywhere we make use of the seed words we end up creating a new variable with the seed words mutated into a single `String` using `join`. Even if we get `Hidden` zeroized correctly, we're still leaving traces of them around every time we make use of them. 
This may be solved with a new seed words struct or could be easy to return the combined string from the source instead of a `Vec<String>`. This rolls into your point about Copy no longer being implemented but it's probably a step in the right direction considering the existing possibility of leaks. I'll also have to look a bit closer around the mnemonic functions and ensure we're not pulling out single words getting left around there as well.

**Comment by AaronFeickert** [comment]
A wrapper design similar to `SafePassword` might do the trick here, similar to what @CjS77 wrote up. You can have it contain a `Hidden<Vec<String>>`, zeroize on drop with a custom `impl Drop` using the mutable reference provided by `Hidden`, expose individual words by reference, and expose the joined word by reference too.

Even though `Hidden` lets you extract the treasure within via `into_inner`, I like the `SafePassword` approach of not doing this. It means callers need to deal with references whose values are safely locked away in a single spot. If there are cases where the caller can't operate by reference, you could either refactor so they can, or implement `into_inner` and then carefully ensure that the caller zeroizes it properly.

**Comment by AaronFeickert** [comment]
I was playing around with a similar idea for another issue dealing with [KDF output](https://github.com/tari-project/tari/issues/4846), and came up with [this fairly generic design](https://github.com/tari-project/tari_utilities/pull/52). A similar non-generic version might be useful for a `SeedWord` struct.

**Comment by stringhandler** [comment]
I prefer making `Hidden` implement `zeroize` to the extra `clandestine` struct (although the name is great). It seems in line with the intention of the struct, and users may expect it to `zeroize`. I was surprised to find out it didn't require it already, for example. 

I do like the idea of a `SeedWords` struct as well, though. Strings should not be passed around containing these words

**Comment by AaronFeickert** [comment]
Just adding `Zeroize` (and zeroize on drop) to `Hidden` still retains the copy and move issues present with any stack allocation, right? Ensuring we also always have a `Box` wrapper makes it easier to assert that zeroization does what it says on the tin.

I also favor the type enforcement present with something like `Clandestine`, which makes it impossible to use data with the same underlying type in the wrong context.

**Comment by AaronFeickert** [comment]
> I prefer making `Hidden` implement `zeroize` to the extra `clandestine` struct (although the name is great). It seems in line with the intention of the struct, and users may expect it to `zeroize`. I was surprised to find out it didn't require it already, for example.
> 
> I do like the idea of a `SeedWords` struct as well, though. Strings should not be passed around containing these words

The [linked PR](https://github.com/tari-project/tari_utilities/pull/52) has been updated to do just this. Now `Hidden` directly supports zeroizing on drop, display/debug hiding, and optional type enforcement. It works seamlessly with `SafePassword` and could likely be used in a similar way for `SeedWords`.

**Comment by stringhandler** [comment]
Closing in favour of #4925

**Pull request closed without merging**